r-package/survlearners/R/surv_rl_f.R

```diff
@@ -0,0 +1,165 @@
+#' }
+#' @return A surv_rl_f object
+#' @export
+surv_rl_f <- function(X, Y, W, D,
+                      t0 = NULL,
+                      k.folds = NULL,
+                      W.hat = NULL,
+                      C.hat = NULL,
+                      new.args.grf.nuisance = list(),
+                      new.args.cf = list(),
+                      cen.fit = "Kaplan-Meier") {
+
+  if (is.null(k.folds)) {
+    k.folds <- floor(max(3, min(10, length(Y) / 4)))
+  }
+
+  if (is.null(W.hat)) {
+    stop("propensity score needs to be supplied")
+  } else if (length(W.hat) == 1) {
+    W.hat <- rep(W.hat, nrow(X))
+  } else if (length(W.hat) != nrow(X)) {
+    stop("W.hat has incorrect length.")
+  }
+
+  args.grf.nuisance <- list(failure.times = NULL,
+                            num.trees = max(50, 2000 / 4),
+                            sample.weights = NULL,
+                            clusters = NULL,
+                            equalize.cluster.weights = FALSE,
+                            sample.fraction = 0.5,
+                            mtry = min(ceiling(sqrt(ncol(X)) + 20), ncol(X)),
+                            min.node.size = 15,
+                            honesty = TRUE,
+                            honesty.fraction = 0.5,
+                            honesty.prune.leaves = TRUE,
+                            alpha = 0.05,
+                            prediction.type = "Nelson-Aalen",
+                            compute.oob.predictions = TRUE,
+                            num.threads = NULL,
+                            seed = runif(1, 0, .Machine$integer.max))
+  args.grf.nuisance[names(new.args.grf.nuisance)] <- new.args.grf.nuisance
+
+  if (is.null(C.hat)) {
+    U <- pmin(Y, t0)                         # truncated follow-up time by t0
+    if (cen.fit == "Kaplan-Meier") {
+      fold.id <- sample(rep(seq(k.folds), length = nrow(X)))
+      C.hat <- rep(NA, length(fold.id))
+      for (z in 1:k.folds) {
+        c.fit <- survival::survfit(survival::Surv(Y[!fold.id == z], 1 - D[!fold.id == z]) ~ 1)
+        kmc <- summary(c.fit, times = U[fold.id == z])
+        C.hat[fold.id == z] <- kmc$surv[match(U[fold.id == z], kmc$time)]
+      }
+    } else if (cen.fit == "survival.forest") {
+      args.grf.nuisance$alpha <- 0.05
+      c.fit <- do.call(grf::survival_forest, c(list(X = cbind(W, X), Y = Y, D = 1 - D), args.grf.nuisance))
+      C.hat <- predict(c.fit, failure.times = U, prediction.times = "time")$predictions
+    }
+  } else {
+    c.fit <- NULL
+  }
+  if (any(C.hat == 0)) {
+    stop("Some or all uncensored probabilities are exactly zeros. Check input variables or consider adjust the time of interest t0.")
+  }
+
+  # CATE function
+  S.t0 <- as.numeric(Y > t0)
+  Y.t0 <- S.t0[D == 1 | Y > t0]
+  W.t0 <- W[D == 1 | Y > t0]
+  W.hat.t0 <- W.hat[D == 1 | Y > t0]
+  X.t0 <- X[D == 1 | Y > t0,, drop = FALSE]
+  C.hat.t0 <- C.hat[D == 1 | Y > t0]
+  sample.weights <- 1 / C.hat.t0
+
```

Comment:
Where is `Y.hat` from the T-learner type GRF Survival Forest?

Reply:
yes